pkg/controller/core.oam.dev/v1alpha2/application/dispatch/dispatch.go

```diff
@@ -0,0 +1,262 @@
+*/
+
+package dispatch
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/pkg/errors"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1beta1"
+	"github.com/oam-dev/kubevela/pkg/utils/apply"
+)
+
+// NewAppManifestsDispatcher creates an AppManifestsDispatcher.
+func NewAppManifestsDispatcher(c client.Client, appRev *v1beta1.ApplicationRevision) *AppManifestsDispatcher {
+	return &AppManifestsDispatcher{
+		c:          c,
+		applicator: apply.NewAPIApplicator(c),
+		appRev:     appRev,
+		gcHandler:  NewGCHandler(c, appRev.Namespace),
+	}
+}
+
+// AppManifestsDispatcher dispatch application manifests into K8s and record the dispatched manifests' references in a
+// resource tracker which is named by a particular rule: name = appRevision's Name + appRevision's namespace.
+// A bundle of manifests to be dispatched MUST come from the given application revision.
+type AppManifestsDispatcher struct {
+	c          client.Client
+	applicator apply.Applicator
+	gcHandler  GarbageCollector
+
+	appRev *v1beta1.ApplicationRevision
+	oldRT  *v1beta1.ResourceTracker
+	skipGC bool
+
+	appRevName string
+	namespace  string
+	newRTName  string
+	newRT      *v1beta1.ResourceTracker
+}
+
+// EnableGC return an AppManifestsDispatcher that always do GC after dispatching resources.
+// GC will calculate diff between the dispatched resouces and ones recorded in the given resource tracker.
+func (a *AppManifestsDispatcher) EnableGC(rt *v1beta1.ResourceTracker) *AppManifestsDispatcher {
+	if rt != nil {
+		a.oldRT = rt.DeepCopy()
+	}
+	return a
+}
+
+// EnableUpgradeAndSkipGC return an AppManifestsDispatcher that skips GC after dispatching resources.
+// For the unchanged resources, dispatcher will update their owner to the newly created resource tracker.
+// It's helpful in a rollout scenario where new revision is going to create a new workload while the old one should not
+// be deleted before rollout is terminated.
+func (a *AppManifestsDispatcher) EnableUpgradeAndSkipGC(rt *v1beta1.ResourceTracker) *AppManifestsDispatcher {
+	if rt != nil {
+		a.oldRT = rt.DeepCopy()
+		a.skipGC = true
+	}
+	return a
+}
+
+// Dispatch apply manifests into k8s and return a resource tracker recording applied manifests' references.
+// If GC is enabled, it will do GC after applying.
+// If 'UpgradeAndSkipGC' is enabled, it will:
+// - create new resources if not exist before
+// - update unchanged resources' owner from the old resource tracker to the new one
+// - skip deleting(GC) any resources
+func (a *AppManifestsDispatcher) Dispatch(ctx context.Context, manifests []*unstructured.Unstructured) (*v1beta1.ResourceTracker, error) {
+	if err := a.completeAndValidate(ctx); err != nil {
+		return nil, err
```

Comment:
I don't see any log or error wraper on it, and the following lines, how do we know where the error from if we need to debug errors

Reply:
I have added logs or wrapped errors in those unexported methods. I think they can provide enough info for debugging.

Reply:
good